### PR TITLE
[ROCm] Add GLM-5.3 k-pool decode-tail maintenance

### PR DESCRIPTION
**Pull request by andyluo7**
## Summary
- add a TileLang prefill kernel that seeds each request rolling BF16 k-pool tail
- add an ordered batched decode kernel for plain and speculative decoding
- compress and write a pool when its closing token arrives
- validate tail ownership, position ordering, padding, and cache-write uniqueness
- add focused ROCm correctness and safety tests

## Dependency
This is stacked on #3250 and includes its commit until that PR lands.

## Validation
- exact head: `bb14b9bdfaacfd36b5201236ebd26fb8b3169240`
- exact-head workflow: https://github.com/tile-ai/tilelang/actions/runs/35427986041
- ROCm 7.2 gfx942 job: https://github.com/tile-ai/tilelang/actions/runs/35427986041/job/105857306286
- ROCm result: `2423 passed, 1852 skipped, 247 warnings`; all six focused GLM decode-tail cases passed
- Quick Lint, CUDA, Metal, CuTeDSL, pre-commit.ci, and CodeRabbit passed on the same head
- local: `python3 -m py_compile`, changed-file pre-commit suite, and `git diff --check`

## Scope
This increment intentionally excludes pool Top-K, pool-to-token expansion, preshuffled framework cache layouts, and vLLM or SGLang integration.

## Summary

- Added ROCm TileLang kernels for GLM-5.3 k-pool tail seeding, ordered decode updates, and pool-closing compression.
- Added BF16 tail-cache maintenance, FP8 cache writes, metadata validation, and cache-write uniqueness checks.
- Added reference implementations, README documentation, and focused ROCm correctness and safety tests.
- Updated FP8 device-selection APIs to accept an explicit device.

## Scope

This change excludes pool Top-K, pool-to-token expansion, preshuffled framework cache layouts, vLLM integration, and SGLang integration.

## Validation

Python compilation, changed-file pre-commit checks, and whitespace validation are complete. Exact-head ROCm runtime validation remains pending CI.

### CODE REVIEW

File: examples/kpool/example_glm53_kpool_decode_tail.py [hunk 1]
```diff
@@ -0,0 +1,569 @@
...
+                            pool_k = T.if_then_else(
+                                pool_slot == tail_offset,
+                                current_k,
+                                historical_k,
+                            )
+                            pooled = pooled + pool_k * probability
+
+                        pooled = T.cast(T.cast(pooled / denom, T.bfloat16), T.float32)
+                        exchange[dim] = pooled
+                        T.sync_threads()
```
[the remark this answers is not in the document]
Comment (andyluo7) [reply examples/kpool/example_glm53_kpool_decode_tail.py:189]: Fixed in `10646888c30b628b98ff1fe97ac5e89fca3d826a`. The pool-close loop now synchronizes the block before overwriting `exchange[dim]`, while retaining the post-write barrier. I also added `test_glm53_kpool_decode_tail_two_pool_closures_reuse_shared_memory` to cover two closures in one launch. Exact-head ROCm gfx942 CI passed, with the full ROCm suite at 2422 passed / 1850 skipped: https://github.com/tile-ai/tilelang/actions/runs/35420297094/job/105836748517

File: examples/kpool/example_glm53_kpool_decode_tail.py [hunk 2]
```diff
@@ -0,0 +1,573 @@
...
+    ahead = torch.full_like(tail_slot_mapping, -1)
+    if tail_slot_mapping.numel() > pool_size:
+        ahead[:-pool_size] = tail_slot_mapping[pool_size:]
+    same_block_ahead = torch.zeros_like(active)
+    nonnegative_ahead = ahead >= 0
+    same_block_ahead[nonnegative_ahead] = ahead[nonnegative_ahead] // pool_size == tail_slot_mapping[nonnegative_ahead] // pool_size
+    seed_mask = active & ~same_block_ahead
+    selected_slots = tail_slot_mapping[seed_mask]
+    if torch.unique(selected_slots).numel() != selected_slots.numel():
+        raise ValueError("selected tail slots must be unique to avoid concurrent writes")
```
[the remark this answers is not in the document]
Comment (andyluo7) [reply examples/kpool/example_glm53_kpool_decode_tail.py]: Fixed in `10646888c30b628b98ff1fe97ac5e89fca3d826a`. Prefill seeding now accepts `cu_seqlens`, launches one program per request, bounds the seed window by that request's end, requires one tail block per request, and rejects shared blocks across active requests. The multi-request boundary/ownership regression passed in exact-head ROCm gfx942 CI: https://github.com/tile-ai/tilelang/actions/runs/35420297094/job/105836748517

File: examples/kpool/example_glm53_kpool_decode_tail.py [hunk 3]
```diff
@@ -0,0 +1,590 @@
...
+            continue
+        if torch.any(request_slots < 0).item() or int(request_slots.max().item()) >= tail_slots:
+            raise ValueError(f"request tail slots must be in [0, {tail_slots})")
+        blocks = request_slots // pool_size
+        if torch.any(blocks != blocks[0]).item():
+            raise ValueError("all prefill tokens for one request must use the same tail block")
+        if request_slots.numel() > 1:
+            phase_step = (request_slots[1:] - request_slots[:-1]).remainder(pool_size)
+            if torch.any(phase_step != 1).item():
+                raise ValueError("prefill tail slots must advance by one modulo pool_size")
```
[the remark this answers is not in the document]
Comment (andyluo7) [reply examples/kpool/example_glm53_kpool_decode_tail.py:322]: Fixed in `10646888c30b628b98ff1fe97ac5e89fca3d826a`. Prefill validation now requires each request's first tail slot to have phase zero, before checking modulo progression. Added `test_glm53_kpool_seed_tail_cache_rejects_nonzero_initial_phase`; it passed in exact-head ROCm gfx942 CI, whose full result was 2422 passed / 1850 skipped: https://github.com/tile-ai/tilelang/actions/runs/35420297094/job/105836748517
